File: VK-Z/Models/VKMusicModels.swift
import Foundation

// MARK: - VK Audio Track
struct VKTrack: Identifiable, Codable, Equatable {
    let id: Int
    let ownerId: Int
    let artist: String
    let title: String
    let duration: Int
    let url: String?
    let albumId: Int?
    let albumTitle: String?
    let genreId: Int?
    let trackCovers: [String]?
    let lyricsId: Int?
    let isExplicit: Bool?
    let isLicensed: Bool?
    let isFeatured: Bool?
    let isHq: Bool?
    
    var formattedDuration: String {
        let minutes = duration / 60
        let seconds = duration % 60
        return String(format: "%d:%02d", minutes, seconds)
    }
    
    var coverUrl: URL? {
        guard let covers = trackCovers, let urlString = covers.last else { return nil }
        return URL(string: urlString)
    }
    
    var audioUrl: URL? {
        guard let urlString = url else { return nil }
        return URL(string: urlString)
    }
    
    enum CodingKeys: String, CodingKey {
        case id
        case ownerId = "owner_id"
        case artist
        case title
        case duration
        case url
        case albumId = "album_id"
        case albumTitle = "album_title"
        case genreId = "genre_id"
        case trackCovers = "track_covers"
        case lyricsId = "lyrics_id"
        case isExplicit = "is_explicit"
        case isLicensed = "is_licensed"
        case isFeatured = "is_featured"
        case isHq = "is_hq"
    }
}

// MARK: - VK Playlist
struct VKPlaylist: Identifiable, Codable, Equatable {
    let id: Int
    let ownerId: Int
    let title: String
    let description: String?
    let tracks: [VKTrack]?
    let trackCount: Int
    let photoUrl: String?
    let accessKey: String?
    let isFavorite: Bool?
    
    var coverUrl: URL? {
        guard let urlString = photoUrl else { return nil }
        return URL(string: urlString)
    }
    
    enum CodingKeys: String, CodingKey {
        case id
        case ownerId = "owner_id"
        case title
        case description
        case tracks
        case trackCount = "track_count"
        case photoUrl = "photo_url"
        case accessKey = "access_key"
        case isFavorite = "is_favorite"
    }
}

// MARK: - VK User
struct VKUser: Codable {
    let id: Int
    let firstName: String
    let lastName: String
    let photoUrl: String?
    let photoUrlBig: String?
    
    var fullName: String { "\(firstName) \(lastName)" }
    var avatarUrl: URL? {
        guard let urlString = photoUrlBig ?? photoUrl else { return nil }
        return URL(string: urlString)
    }
    
    enum CodingKeys: String, CodingKey {
        case id
        case firstName = "first_name"
        case lastName = "last_name"
        case photoUrl = "photo_url"
        case photoUrlBig = "photo_url_big"
    }
}

// MARK: - VK Audio Section (Catalog)
struct VKCatalogSection: Identifiable, Codable {
    let id: String
    let title: String
    let type: String
    let playlists: [VKPlaylist]?
    let tracks: [VKTrack]?
}

// MARK: - API Response Wrappers
struct VKApiResponse<T: Codable>: Codable {
    let response: T?
    let error: VKApiError?
}

struct VKApiError: Codable {
    let errorCode: Int
    let errorMsg: String
    
    enum CodingKeys: String, CodingKey {
        case errorCode = "error_code"
        case errorMsg = "error_msg"
    }
}

struct VKMusicResponse: Codable {
    let count: Int
    let items: [VKTrack]
}

struct VKPlaylistsResponse: Codable {
    let count: Int
    let items: [VKPlaylist]
}

struct VKSearchResponse: Codable {
    let count: Int
    let items: [VKTrack]
}

// MARK: - VK Mix (умный плейлист/радио)
struct VKMix: Identifiable, Codable, Equatable {
    let id: String
    let title: String
    let subtitle: String?
    let description: String?
    let icon: String?
    let coverUrl: String?
    let color: String?
    let tracks: [VKTrack]?
    let trackCount: Int?
    let isMix: Bool?
    let isFollowed: Bool?
    let artists: [String]?
    
    var displayCover: URL? {
        if let urlString = coverUrl { return URL(string: urlString) }
        return tracks?.first?.coverUrl
    }
    
    enum CodingKeys: String, CodingKey {
        case id
        case title
        case subtitle
        case description
        case icon
        case coverUrl = "cover_url"
        case color
        case tracks
        case trackCount = "track_count"
        case isMix = "is_mix"
        case isFollowed = "is_followed"
        case artists
    }
}

struct VKMixResponse: Codable {
    let count: Int
    let items: [VKMix]
}

struct VKMixTracksResponse: Codable {
    let count: Int
    let items: [VKTrack]
}

// MARK: - Player State
enum PlayerRepeatMode: String, Codable {
    case off
    case all
    case one
}

enum PlayerState {
    case stopped
    case playing
    case paused
    case loading
    case error(String)
}

// MARK: - Auth State
enum AuthState {
    case unauthenticated
    case authenticating
    case authenticated(token: String, userId: Int)
    case error(String)
}

File: windows/VKZ/Models/VKMusicModels.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Newtonsoft.Json;

namespace VKZ.Models
{
    public class VKTrack
    {
        [JsonProperty("id")]
        public long Id { get; set; }

        [JsonProperty("owner_id")]
        public long OwnerId { get; set; }

        [JsonProperty("title")]
        public string Title { get; set; } = string.Empty;

        [JsonProperty("artist")]
        public string Artist { get; set; } = string.Empty;

        [JsonProperty("duration")]
        public int Duration { get; set; }

        [JsonProperty("url")]
        public string? Url { get; set; }

        [JsonProperty("track_code")]
        public string? TrackCode { get; set; }

        [JsonProperty("access_key")]
        public string? AccessKey { get; set; }

        [JsonProperty("album_id")]
        public long? AlbumId { get; set; }

        [JsonIgnore]
        public string FormattedDuration
        {
            get
            {
                var ts = TimeSpan.FromSeconds(Duration);
                return ts.Hours > 0
                    ? $"{ts.Hours}:{ts.Minutes:D2}:{ts.Seconds:D2}"
                    : $"{ts.Minutes}:{ts.Seconds:D2}";
            }
        }

        [JsonIgnore]
        public string? CoverUrl
        {
            get
            {
                if (AlbumId == null || OwnerId == 0) return null;
                return $"https://music.vk.com/api/audio.getCover?owner_id={OwnerId}&album_id={AlbumId}";
            }
        }

        [JsonIgnore]
        public string AudioUrl => Url ?? string.Empty;

        [JsonIgnore]
        public string FullId => $"{OwnerId}_{Id}";

        public override bool Equals(object? obj) =>
            obj is VKTrack other && OwnerId == other.OwnerId && Id == other.Id;

        public override int GetHashCode() => HashCode.Combine(OwnerId, Id);
    }

    public class VKPlaylist
    {
        [JsonProperty("id")]
        public long Id { get; set; }

        [JsonProperty("owner_id")]
        public long OwnerId { get; set; }

        [JsonProperty("title")]
        public string Title { get; set; } = string.Empty;

        [JsonProperty("description")]
        public string? Description { get; set; }

        [JsonProperty("access_key")]
        public string? AccessKey { get; set; }

        [JsonProperty("count")]
        public int Count { get; set; }

        [JsonProperty("photo")]
        public VKPhoto? Photo { get; set; }

        [JsonIgnore]
        public string? CoverUrl => Photo?.Photo600 ?? Photo?.Photo300;

        [JsonIgnore]
        public string FullId => $"{OwnerId}_{Id}";
    }

    public class VKPhoto
    {
        [JsonProperty("photo_300")]
        public string? Photo300 { get; set; }

        [JsonProperty("photo_600")]
        public string? Photo600 { get; set; }

        [JsonProperty("photo_1200")]
        public string? Photo1200 { get; set; }
    }

    public class VKUser
    {
        [JsonProperty("id")]
        public long Id { get; set; }

        [JsonProperty("first_name")]
        public string FirstName { get; set; } = string.Empty;

        [JsonProperty("last_name")]
        public string LastName { get; set; } = string.Empty;

        [JsonProperty("photo_200")]
        public string? Photo200 { get; set; }

        [JsonIgnore]
        public string FullName => $"{FirstName} {LastName}";

        [JsonIgnore]
        public string? AvatarUrl => Photo200;
    }

    public class VKMix
    {
        [JsonProperty("id")]
        public string Id { get; set; } = string.Empty;

        [JsonProperty("title")]
        public string Title { get; set; } = string.Empty;

        [JsonProperty("description")]
        public string? Description { get; set; }

        [JsonProperty("cover_url")]
        public string? CoverUrl { get; set; }

        [JsonProperty("color")]
        public string? Color { get; set; }

        [JsonProperty("track_count")]
        public int TrackCount { get; set; }
    }

    public class VKApiResponse<T>
    {
        [JsonProperty("response")]
        public T? Response { get; set; }

        [JsonProperty("error")]
        public VKApiError? Error { get; set; }
    }

    public class VKApiError
    {
        [JsonProperty("error_code")]
        public int ErrorCode { get; set; }

        [JsonProperty("error_msg")]
        public string ErrorMsg { get; set; } = string.Empty;
    }

    public class VKMusicResponse
    {
        [JsonProperty("items")]
        public List<VKTrack> Items { get; set; } = new();
    }

    public class VKPlaylistsResponse
    {
        [JsonProperty("items")]
        public List<VKPlaylist> Items { get; set; } = new();
    }

    public class VKSearchResponse
    {
        [JsonProperty("items")]
        public List<VKTrack> Items { get; set; } = new();
    }

    public class VKMixResponse
    {
        [JsonProperty("items")]
        public List<VKMix> Items { get; set; } = new();
    }

    public class VKMixTracksResponse
    {
        [JsonProperty("items")]
        public List<VKTrack> Items { get; set; } = new();
    }

    public enum PlayerRepeatMode
    {
        None,
        One,
        All
    }

    public enum PlayerState
    {
        Idle,
        Playing,
        Paused,
        Loading
    }

    public enum AuthState
    {
        Unauthorized,
        Authorizing,
        Authorized,
        Error
    }
}